pkg/ccl/sqlproxyccl/balancer/conn_tracker.go

```diff
@@ -199,16 +278,58 @@ func (e *tenantEntry) getConnsMap() map[string][]ConnectionHandle {
 	// Note that holding onto the lock for a long time will affect latency of
 	// incoming connections.
 	conns := make(map[string][]ConnectionHandle)
-	for sa := range e.mu.assignments {
-		conns[sa.Addr()] = append(conns[sa.Addr()], sa.Owner())
+	for sa := range e.assignments.active {
+		if sa.Owner() != nil {
+			conns[sa.Addr()] = append(conns[sa.Addr()], sa.Owner())
+		}
+	}
+	for sa := range e.assignments.idle {
+		if sa.Owner() != nil {
+			conns[sa.Addr()] = append(conns[sa.Addr()], sa.Owner())
+		}
 	}
 	return conns
 }
 
-// getConnsCount returns the number of connections associated with the tenant.
-func (e *tenantEntry) getConnsCount() int {
+// assignmentsCount returns the number of assignments associated with the tenant.
+func (e *tenantEntry) assignmentsCount() int {
 	e.mu.Lock()
 	defer e.mu.Unlock()
+	return len(e.assignments.active) + len(e.assignments.idle)
+}
+
+// refreshPartitions updates the partitions of the server assignments associated
+// with this tenant. Assignments with active owners will be moved to the active
+// partition, and a similar argument can be made for the idle partition with
+// idle owners.
+func (e *tenantEntry) refreshPartitions() {
+	// moveTo moves the given server assignment sa from the `from` partition to
+	// the `to` partition. If the assignment is no longer present in the `from`
+	// partition, this will be a no-op.
+	moveTo := func(
+		from map[*ServerAssignment]struct{},
+		to map[*ServerAssignment]struct{},
+		sa *ServerAssignment,
+	) {
+		e.mu.Lock()
+		defer e.mu.Unlock()
+		// Assignment is no longer present.
+		if _, ok := from[sa]; !ok {
+			return
+		}
+		delete(from, sa)
+		to[sa] = struct{}{}
+	}
 
-	return len(e.mu.assignments)
+	activeList, idleList := e.listAssignments()
+	for _, sa := range activeList {
+		if sa.Owner() != nil && sa.Owner().IsIdle() {
+			moveTo(e.assignments.active, e.assignments.idle, sa)
```

Comment:
The way moveTo is called here violates a common convention with go locking. The assignments field is protected by the mutex and this retrieves the .active and .idle maps without holding the lock. This would result in a race condition if someone acquired e.mu and overwrote one of the maps (e.g. e.mu.asignemnts.idle = clone(e.mu.assignments.idle)).

I like the api of moveTo, but replacing it with something like this allows us to keep to the go locking conventions:
```
func setActive(active bool, sa *ServerAssigment) {
    e.mu.Lock()
    defer e.mu.unlock()
    if active && e.assignments.inactive[sa] {
        delete(e.assignments.inactive[sa])     
        e.assignments.active[sa] = true
    } 
    if !active && e.assignments.active[sa] {
        delete(e.assignments.active)
        e.assignments.inactive[sa] = true
    }
}
```